README.md
# URL Mangler

VS Code extension that formats highlighted URLs in the text editor.


## Features

![URL Mangler demonstration](doc/demo.gif)

**Recognises URLs for**:

* Confluence
* JIRA
* And can be used on URLs form other sites. YMMV.

**Detection of**:

* URL fragments which link to a section within the page. The section title is extracted as part of the display name.

**Extracts display names to the following formats**:

|Format|Pattern|Comments|
|------|-------|--------|
|Display name only|``{display name}``|Extracts just the display name only without the URL|
|HTML href|``<a href="{display name}">{url}</a>``|Extracts as a HTML ``<a>`` anchor tag. Note this is plain text, not text/html. See [Known Issues](#known-issues) section below. |
|JIRA text syntax|``[{display name}\|{url}]``|Extracts to JIRA's text syntax|
|Markdown|``[{display name}]({url})``|Extracts to Markdown syntax|


## Known Issues

### HTML href output

The HTML href output format updates the clipboard with ``text/plain`` content. This should be ``text/html`` but [VS Code's Clipboard API](https://code.visualstudio.com/api/references/vscode-api#Clipboard) only allows you to write plain text. This could change if issue [141704](https://github.com/microsoft/vscode/issues/141704) is accepted although [163240](https://github.com/microsoft/vscode/issues/163240) was closed because of [incomplete browser compatibility for Clipboard#write](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/write).

### Extension installation

Currently this isn't deployed to the VS Code Marketplace.


## Release Notes

### 0.0.1 - 2023-07-15

Initial release


CHANGELOG.md
# Change Log

All notable changes to the "urlmanglervscodeextension" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/test/suite/extension.test.ts
import * as assert from 'assert';
import { describe, it } from 'mocha';

// You can import and use all API from the 'vscode' module
// as well as import your extension to test it
import * as vscode from 'vscode';

import { OutputFormat, extract, formatDisplayName } from '../../mangler';

suite('URL Mangler Extension Test Suite', () => {
	vscode.window.showInformationMessage('Start all tests.');

	describe('Extract', function () {
		it('Not a URL', function () {
			assert.strictEqual('', extract('foo'));
		});
		it('Confluence-like', function () {
			// The real https://confluence.atlassian.com/confkb/the-differences-between-various-url-formats-for-a-confluence-page-278692715.html has a URL extension and page ID in the URL
			assert.strictEqual('The Differences Between Various Url Formats For A Confluence Page', extract('https://confluence.atlassian.com/confkb/the-differences-between-various-url-formats-for-a-confluence-page'));
		});
		it('Confluence-like with simple fragment', function () {
			assert.strictEqual('The Differences Between Various URL Formats for a Confluence Page#Background', extract('https://confluence.atlassian.com/confkb/the-differences-between-various-url-formats-for-a-confluence-page#TheDifferencesBetweenVariousURLFormatsforaConfluencePage-Background'));
		});
		it('JIRA', function () {
			assert.strictEqual('CONFCLOUD-74340', extract('https://jira.atlassian.com/browse/CONFCLOUD-74340'));
			assert.strictEqual('CONFCLOUD-74340', extract('https://jira.atlassian.com/browse/CONFCLOUD-74340?filter=-5'));
		});
	});

	describe('Extract with output formatting', function () {
		it('href', function () {
			assert.strictEqual('<a href="https://confluence.atlassian.com/confkb/the-differences-between-various-url-formats-for-a-confluence-page#TheDifferencesBetweenVariousURLFormatsforaConfluencePage-Background">The Differences Between Various URL Formats for a Confluence Page#Background</a>', formatDisplayName('https://confluence.atlassian.com/confkb/the-differences-between-various-url-formats-for-a-confluence-page#TheDifferencesBetweenVariousURLFormatsforaConfluencePage-Background', 'The Differences Between Various URL Formats for a Confluence Page#Background', OutputFormat.href));
		});
		it('Jira', function () {
			assert.strictEqual('[The Differences Between Various URL Formats for a Confluence Page#Background|https://confluence.atlassian.com/confkb/the-differences-between-various-url-formats-for-a-confluence-page#TheDifferencesBetweenVariousURLFormatsforaConfluencePage-Background]', formatDisplayName('https://confluence.atlassian.com/confkb/the-differences-between-various-url-formats-for-a-confluence-page#TheDifferencesBetweenVariousURLFormatsforaConfluencePage-Background', 'The Differences Between Various URL Formats for a Confluence Page#Background', OutputFormat.jira));
		});
		it('Markdown', function () {
			assert.strictEqual('[The Differences Between Various URL Formats for a Confluence Page#Background](https://confluence.atlassian.com/confkb/the-differences-between-various-url-formats-for-a-confluence-page#TheDifferencesBetweenVariousURLFormatsforaConfluencePage-Background)', formatDisplayName('https://confluence.atlassian.com/confkb/the-differences-between-various-url-formats-for-a-confluence-page#TheDifferencesBetweenVariousURLFormatsforaConfluencePage-Background', 'The Differences Between Various URL Formats for a Confluence Page#Background', OutputFormat.markdown));
		});
	});
});


src/mangler.ts
export enum OutputFormat {
	displayNameOnly, // Just the display name only
	href, // A href with extracted display name and original URL
	jira, // A link in JIRA's raw text format [<display name>|URL]
	markdown // A link in Markdown format [<display name>](URL)]
}

const REGEX_CHARS_TO_REPLACE = /[-+_]/g;

/**
 * For the given the URL and display name to the specified output format.
 * @param urlString the URL the displayName was extracted from
 * @param displayName the display name for the URL
 * @param outputFormat the desired output format. Defaults to OutputFormat.displayNameOnly
 * @returns a formatted string
 */
export function formatDisplayName(urlString: string, displayName: string, outputFormat: OutputFormat = OutputFormat.displayNameOnly): string {
	switch (outputFormat) {
		// TODO need some escaping here if the display name has special characters
		case OutputFormat.href:
			return `<a href="${urlString}">${displayName}</a>`;
		case OutputFormat.jira:
			return `[${displayName}|${urlString}]`;
		case OutputFormat.markdown:
			return `[${displayName}](${urlString})`;
	}
	return displayName;
}

/**
 * Extract the display from the the given URL
 * @param urlString the URL
 * @returns the display name or the empty string if a display name can't be extracted
 */
export function extract(urlString: string): string {
	let url: URL;
	try {
		url = new URL(urlString);
	} catch (err) {
		// Not a URL
		return '';
	}

	const lastSlash: number = url.pathname.lastIndexOf('/');
	let title: string = url.pathname.substring(lastSlash + 1);
	let hash = url.hash;

	// JIRA handling: preserve '-'
	if (url.hostname.toLowerCase().indexOf('jira') >= 0) {
		return title;
	}

	// Confluence handling: insert spaces and handle fragments
	let result = title.replaceAll(REGEX_CHARS_TO_REPLACE, ' ');
	if (hash) {
		hash = hash.substring(1);
		// The hash gives us a big hint as how to format the case (but not the insertion of spaces) of the display name
		let titleNoSpaces = title.replaceAll(REGEX_CHARS_TO_REPLACE, '');
		if (hash.toLowerCase().startsWith(titleNoSpaces.toLowerCase())) {
			result = toFragmentCase(result, hash);
			hash = hash.substring(titleNoSpaces.length + 1);
		} else {
			// Can't use the hash to derive the case, just use sentence case
			result = toSentenceTitleCase(result);
		}

		// Append fragment
		hash = hash.replaceAll(REGEX_CHARS_TO_REPLACE, ' ').trim();
		result += '#' + toSentenceTitleCase(hash);

	} else {
		result = toSentenceTitleCase(title.replaceAll(REGEX_CHARS_TO_REPLACE, ' '));
	}

	return result;
}

/**
 * Converts a space separated sentence phrase and changes it to title case (i.e. the first letter of each word is upper case). The case of other letters is not adjusted to handle cases where upper casing has already been applied.
 * @param text the sentence
 * @returns the sentence in title case
 */
function toSentenceTitleCase(text: string): string {
	let words: string[] = text.split(' ');
	words = words.map(function (word) {
		return word.charAt(0).toUpperCase() + word.slice(1);
	});
	return words.join(' ');
}

/**
 * Converts the given sentence using the same case as given fragment. This is useful in sites such as Confluence where the path is always lower case but the fragment has the case of the title of the page.
 * @param text the sentence with spaces (this function doesn't add new spaces)
 * @param fragment the fragment which must not have space. Can be longer than the sentence (i.e. has a genuine fragment at the end)
 * @returns the sentence using the fragment's case
 */
function toFragmentCase(text: string, fragment: string): string {
	let result: string[] = [];
	let j: number = 0;
	for (let i: number = 0; i < text.length; i++) {
		let c = text.charAt(i);
		if (c === ' ') {
			result.push(' '); // Advance through the input text that has spaces but not the fragment which doesn't have spaces
		} else {
			result.push(fragment.charAt(j));
			j++;
		}
	}
	return result.join('');
}

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { OutputFormat, extract, formatDisplayName } from './mangler';

// This method is called when your extension is activated
// Your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	//console.log('Congratulations, your extension "urlmangler" is now active!');

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json
	let disposable = vscode.commands.registerCommand('urlmangler.extract', async () => {
		// The code you place here will be executed every time your command is executed
		// Display a message box to the user
		const editor = vscode.window.activeTextEditor;
		const selection = editor?.selection;
		if (selection && !selection.isEmpty) {
			const selectionRange = new vscode.Range(selection.start.line, selection.start.character, selection.end.line, selection.end.character);
			const highlighted = editor.document.getText(selectionRange);
			const displayName = extract(highlighted);
			if (displayName === '') {
				vscode.window.showWarningMessage(`Can't mangle the URL '${highlighted}', it doesn't appear to be a URL`);
			} else {
				const item = await vscode.window.showQuickPick(['Display name only', 'href', 'Jira text', 'Markdown'], { title: 'Select an output format' });
				let outputFormat;
				switch (item) {
					case 'href':
						outputFormat = OutputFormat.href;
						break;
					case 'Jira text':
						outputFormat = OutputFormat.jira;
						break;
					case 'Markdown':
						outputFormat = OutputFormat.markdown;
						break;
					default:
						outputFormat = OutputFormat.displayNameOnly;
				}
				const result = formatDisplayName(highlighted, displayName, outputFormat);

				// TODO this only lets you update the clipboard with text/plain, not text/html which an href would need. See:
				// https://github.com/microsoft/vscode/issues/163240 - closed as not planned since Clipboard API support isn't there
				// https://github.com/microsoft/vscode/issues/141704 - still open
				vscode.env.clipboard.writeText(result);

				vscode.window.showInformationMessage('Mangled URL to clipboard: ' + result);
			}
		} else {
			vscode.window.showInformationMessage('Select a URL to mangle');
		}
	});

	context.subscriptions.push(disposable);
}

// This method is called when your extension is deactivated
export function deactivate() { }
